ctapipe/image/time_next_neighbor_cleaning.py

```diff
@@ -0,0 +1,742 @@
+        self._camera_name = None
+        self._IPR = None
+        self.__neighbor_group = None
+        self._multiplicity = None
+
+    @property
+    def _geometry(self):
+        """
+        Get or set the current `_geometry. Setting it to the new value also
+        sets `_camera_name` and selects the corresponding `_IPR` from `IPR_dict`.
+        """
+        return self.__geometry
+
+    @_geometry.setter
+    def _geometry(self, geometry):
+        self.__geometry = geometry
+        self._camera_name = geometry.camera_name
+        try:
+            self._IPR = self.IPR_dict[geometry.camera_name]
+        except KeyError as e:
+            raise KeyError(
+                f"No IPR graph for {geometry.camera_name} in" f" self.IPR_dict."
+            ) from e
+
+    @property
+    def _neighbor_group(self):
+        """
+        Get or set the current _neighbor_group. Setting the _neighbor_group to
+        a new value also changes `_multiplicity' to the corresponding value.
+        """
+        return self.__neighbor_group
+
+    @_neighbor_group.setter
+    def _neighbor_group(self, nfold):
+        self.__neighbor_group = nfold
+        self._multiplicity = self._multiplicity_dict[nfold]
+
+    def load_individual_pixel_rates(self, file):
+        """
+        Load the IPR graphs from a json file. The dictionaries should contain
+        one dictionary for each camera_id with the pixel rates and charges.
+
+        Parameters
+        ----------
+        file: string
+            json file in which the IPR graphs is stored
+        """
+        with open(file, "r") as f:
+            self.IPR_dict = json.load(f)
+
+    def dump_individual_pixel_rates(self, file):
+        """
+        Dump the IPR graph to file using json.
+
+        Parameters
+        ----------
+        file: string
+
+        """
+        with open(file, "w") as f:
+            json.dump(self.IPR_dict, f, sort_keys=True, indent=4)
+
+    def fill_individual_pixel_rate(self, geo, image, trace_length, charge_steps):
+        """
+        Count the number of pixels in images about a given charge threshold
+        and fill dictionary with individual pixel noise rate (IPR).
+
+        Parameters
+        ----------
+        geo: ctapipe.instrument.CameraGeometry
+        image: numpy.ndarray
+        trace_length: astropy.units.Quantity
+            total length of the trace, e.g. number of samples times sample_time
+        charge_steps: numpy.ndarray
+            array with the charge steps used to calculate the IPR graphs.
+        """
+
+        self.IPR_dict = self.IPR_dict if self.IPR_dict else {}
+
+        if geo.camera_name not in self.IPR_dict:
+            self.IPR_dict[geo.camera_name] = {
+                "counter": 0,
+                "charge": charge_steps,
+                "npix": np.zeros(charge_steps.shape[0]),
+                "rate": np.zeros(charge_steps.shape[0]),
+                "rate_err": np.zeros(charge_steps.shape[0]),
+            }
```

Comment:
See previous comment. I think using a Table format for this, rather than a  dict would be better.